frame/evm/src/lib.rs

```diff
@@ -512,9 +556,21 @@ pub mod pallet {
 	#[pallet::storage]
 	pub type AccountCodes<T: Config> = StorageMap<_, Blake2_128Concat, H160, Vec<u8>, ValueQuery>;
 
+	#[pallet::storage]
+	pub type AccountCodesAccessed<T: Config> =
```

Comment:
Yeah I think we tried to over-optimize here and it was a really bad idea

Reply:
Wait, I'm not agree with that.
I think it's more of a problem to account too much pov, as this will severely limit the size of blocks in the event of spike usage of some contracts.
Moreover, it's not a problem that certain transactions cost less depending on previous transactions; on the contrary, it's beneficial for the network: those who can wait will arbitrate so that their transactions cost less, thus optimizing the blockspace.

Art least, it's a design choice that must be made for each project, it's not up to frontier to impose it.

Reply:
> They also don't have any incentive to optimize the blocks themselves (because it will lead to them receiving less fees)

It's the opposite. You seem to be assuming that the collators/validators are a small group that collude with each other, but that's not the case (or else we're talking about a centralized chain, in which case there's no point in making a blockchain).

The reality is that when a collator/validator is lucky enough to be eligible on a slot, his interest is to maximize his winnings on that slot, because he doesn't know when he'll be eligible again. To do this, it must include as many transactions as possible, it's only in cases where there are more ready transactions than available space in the block is the collator encouraged to prioritize cold transactions, but outside spike activity periods, collators have an incentive to include all transactions as soon as they can.

Furthermore, in practice, collators cannot recognize whether a transaction is going to be cold or warn without executing it.
The gas estimation always considers the worst case (cold), so the gas limit and therefore the transaction priority will be the same whether the transaction is cold or warn. It's just that in some cases there will be a refund, which isn't a problem for the collator because the refund means that the transaction has consumed fewer resources than expected, so there's more room than expected to include other transactions.

Reply:
> Furthermore, in practice, collators cannot recognize whether a transaction is going to be cold or warn without executing it.

I think this can easily be predicted by looking at the contract being called, e.g. all calls to the same DEX are likely to touch many of the same storage items.

Reply:
I agree with @sorpaas here (I actually brought this up earlier in this PR), but I think @librelois makes good points as well. I'd like to propose something in between:

* Enforce an optimized per-block limit on proof size (allow for full blocks)
* Provide a partial refund to the txn for warm proof changes

This would give block authors the proper incentive to maximize block proof space while splitting some of this benefit with users.

The refund could be configurable, allowing an implementation to effectively disable it in favor of either users or authors.

Reply:
I agree with the fact of using a "per-transaction" computation, not really because of risks or economic impact, but simply to ensure keeping it as simple as possible for the first iteration. Having to explain users that their gas cost would be dependent of previous transactions in the block would not be trivial.

Once we gather more feedback and metrics, if the optimization of using it per block is valuable, we can redesign this solution a bit to support both cases and offer the possibility to each blockchain to choose.